### PR TITLE
feat(functions/utils) add the tribe_without_filters function

### PR DESCRIPTION
**Pull request by lucatume**
Issue: https://moderntribe.atlassian.net/browse/TEC-3579

[Screencast](https://drive.google.com/open?id=1CUDUUP10O2N1n9-bm47NTcdpuaE-FXZX&authuser=luca%40tri.be&usp=drive_fs)

This PR adds the `tribe_without_filters` function to run a callback suspending (detaching and re-attaching) a set of filters and actions.
The function is just a cleaner and tested approach to the pattern:

```php
// Detach this to avoid issues.
remove_filter( 'some_filter', 'some_function' );
$value = my_function();
// Re-attach the function to the filter.
add_filter( 'some_filter', 'some_function' );
```

With the function:

```php
$value = tribe_without_filters( ['some_filter'], 'my_function' );
```

### CODE REVIEW

File: tests/wpunit/Tribe/functions/utilsTest.php [hunk 1]
```diff
@@ -254,4 +309,143 @@ public function tribe_get_query_var_data_set() {
...
+	 *
+	 */
+	public function provider_get_unsuccessful_class_instance() {
+		yield 'empty string' => [ '' ];
+		yield 'string that is neither an existing class name nor slug' => [ 'ABC_123_XYZ' ];
+		yield 'neither an object nor a string' => [ [] ];
+	}
+
+	/**
+	 * It should allow running a callback detaching filters
```
Comment (Camwyn) [inline tests/wpunit/Tribe/functions/utilsTest.php:365]:
```suggestion
	 * It should allow running a callback detaching filters.
```

File: tests/wpunit/Tribe/functions/utilsTest.php [hunk 2]
```diff
@@ -254,4 +309,143 @@ public function tribe_get_query_var_data_set() {
...
+			return apply_filters( 'test_filter', 23 );
+		};
+
+		$value = tribe_without_filters( [ 'test_filter' ], $callback );
+
+		$this->assertEquals( 23, $value );
+	}
+
+	/**
+	 * It should allow running a callback detaching actions
```
Comment (Camwyn) [inline tests/wpunit/Tribe/functions/utilsTest.php:381]:
```suggestion
	 * It should allow running a callback detaching actions.
```

File: tests/wpunit/Tribe/functions/utilsTest.php [hunk 3]
```diff
@@ -254,4 +309,143 @@ public function tribe_get_query_var_data_set() {
...
+			return 23;
+		};
+
+		$value = tribe_without_filters( [ 'test_action' ], $callback );
+
+		$this->assertEquals( 23, $value );
+	}
+
+	/**
+	 * It should allow running a callback detaching filters and actions
```
Comment (Camwyn) [inline tests/wpunit/Tribe/functions/utilsTest.php:401]:
```suggestion
	 * It should allow running a callback detaching filters and actions.
```

File: tests/wpunit/Tribe/functions/utilsTest.php [hunk 4]
```diff
@@ -254,4 +309,143 @@ public function tribe_get_query_var_data_set() {
...
+			return apply_filters( 'test_filter', 23 );
+		};
+
+		$value = tribe_without_filters( [ 'test_action', 'test_filter' ], $callback );
+
+		$this->assertEquals( 23, $value );
+	}
+
+	/**
+	 * It should not detach filters that are not in the list of filters to suspend
```
Comment (Camwyn) [inline tests/wpunit/Tribe/functions/utilsTest.php:424]:
```suggestion
	 * It should not detach filters that are not in the list of filters to suspend.
```

**Comment by sc0ttkclark** [comment]
I'd much rather this function not totally disable all hooks on that filter/action and focus on the function passed in. Maybe make that optional?

**Review by sc0ttkclark — CHANGES_REQUESTED** [review]
Looks great, but maybe make the removing of all hooks on that filter/action optional and default to just removing it for the current passed callback.

**Comment by lucatume** [comment]
> I'd much rather this function not totally disable all hooks on that filter/action and focus on the function passed in. Maybe make that optional?

The function passed in is not the one that is hooked to a filter or action, but the one that will be executed suspending all filters and actions listed in the first argument.

**Review by Camwyn — APPROVED** [review]

**Comment by sc0ttkclark** [comment]
Oh I read this as intending to prevent recursion loops for a callback hooked into a filter/action that will run again. My bad there, this is good.

**Review by sc0ttkclark — APPROVED** [review]

**Pull request merged**